File: ESP32-Environment/components/mesh_common/src/topology_graph.c
/**
 * @file topology_graph.c
 * @brief See topology_graph.h. Pure C - builds on the host for tests
 *        (tools/topology_graph_host_test.c).
 */

#include "topology_graph.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ── helpers ─────────────────────────────────────────────────────────────── */

static uint64_t mac48(const uint8_t *m)
{
    uint64_t v = 0;
    for (int i = 0; i < 6; i++) {
        v = (v << 8) | m[i];
    }
    return v;
}

static bool mac_is_zero(const uint8_t *m)
{
    for (int i = 0; i < 6; i++) {
        if (m[i] != 0) {
            return false;
        }
    }
    return true;
}

static void mac_str(const uint8_t *m, char *out)
{
    snprintf(out, 18, "%02x:%02x:%02x:%02x:%02x:%02x",
             m[0], m[1], m[2], m[3], m[4], m[5]);
}

/* A parent link is the parent's SoftAP BSSID, which on the ESP32 is its STA
 * MAC + 1. An exact STA match is accepted too, so callers that already hold
 * STA MACs work unchanged. */
static int resolve(const topo_node_t *nodes, size_t n, const uint8_t *link)
{
    if (mac_is_zero(link)) {
        return -1;
    }
    uint64_t p = mac48(link);
    for (size_t j = 0; j < n; j++) {
        uint64_t v = mac48(nodes[j].mac);
        if (v + 1 == p || v == p) {
            return (int)j;
        }
    }
    return -1;
}

/* Nodes reachable from @p start over child lists; fills layer[] if given. */
static int bfs_from(const topo_graph_t *g, int start, int *queue, int *layer)
{
    int head = 0, tail = 0, count = 0;
    queue[tail++] = start;
    if (layer) {
        layer[start] = 1;
    }
    while (head < tail) {
        int u = queue[head++];
        count++;
        for (int c = g->first_child[u]; c >= 0; c = g->next_sibling[c]) {
            if (layer) {
                layer[c] = layer[u] + 1;
            }
            queue[tail++] = c;
        }
    }
    return count;
}

/* ── build ───────────────────────────────────────────────────────────────── */

bool topo_build(topo_graph_t *g, const topo_node_t *nodes, size_t n)
{
    memset(g, 0, sizeof(*g));
    g->n = n;
    g->root = -1;
    if (n == 0) {
        return true;
    }

    g->parent       = malloc(n * sizeof(int));
    g->layer        = calloc(n, sizeof(int));
    g->child_count  = calloc(n, sizeof(int));
    g->first_child  = malloc(n * sizeof(int));
    g->next_sibling = malloc(n * sizeof(int));
    int *queue      = malloc(n * sizeof(int));
    if (!g->parent || !g->layer || !g->child_count || !g->first_child ||
        !g->next_sibling || !queue) {
        free(queue);
        topo_free(g);
        return false;
    }

    for (size_t i = 0; i < n; i++) {
        g->first_child[i]  = -1;
        g->next_sibling[i] = -1;
        int p = resolve(nodes, n, nodes[i].parent);
        if (p == (int)i) {
            p = -1;   /* a node can't parent itself; treat as a broken link */
            g->cycle = true;
        }
        if (p < 0 && !mac_is_zero(nodes[i].parent)) {
            g->unresolved++;
        }
        g->parent[i] = p;
    }

    /* Child lists in ascending index order: prepend while walking backwards. */
    for (size_t k = n; k-- > 0;) {
        int p = g->parent[k];
        if (p >= 0) {
            g->next_sibling[k] = g->first_child[p];
            g->first_child[p]  = (int)k;
            g->child_count[p]++;
        }
    }

    /* A parent chain that never reaches a parentless node is a cycle. */
    for (size_t i = 0; i < n && !g->cycle; i++) {
        int cur = (int)i;
        size_t steps = 0;
        while (cur >= 0 && steps <= n) {
            cur = g->parent[cur];
            steps++;
        }
        if (cur >= 0) {
            g->cycle = true;
        }
    }

    /* Root = the parentless node that carries the most of the mesh. Nothing
     * about its ID or position is assumed; a node that hasn't joined yet is
     * also parentless but reaches only itself. A node whose parent exists but
     * isn't in the set (not reported yet) is not a root - it only competes if
     * no genuinely parentless node exists. */
    int best = -1, best_count = 0;
    for (int allow_missing = 0; allow_missing <= 1 && best < 0; allow_missing++) {
        for (size_t i = 0; i < n; i++) {
            if (g->parent[i] >= 0 ||
                (!allow_missing && !mac_is_zero(nodes[i].parent))) {
                continue;
            }
            int c = bfs_from(g, (int)i, queue, NULL);
            if (c > best_count) {
                best = (int)i;
                best_count = c;
            }
        }
    }
    g->root = best;

    if (best >= 0) {
        g->reachable = bfs_from(g, best, queue, g->layer);
        for (size_t i = 0; i < n; i++) {
            if (g->layer[i] > g->max_layer) {
                g->max_layer = g->layer[i];
            }
        }
    }
    free(queue);
    return true;
}

void topo_free(topo_graph_t *g)
{
    free(g->parent);
    free(g->layer);
    free(g->child_count);
    free(g->first_child);
    free(g->next_sibling);
    memset(g, 0, sizeof(*g));
    g->root = -1;
}

/* ── walk ────────────────────────────────────────────────────────────────── */

void topo_walk(const topo_graph_t *g,
               void (*cb)(void *ctx, int idx, int depth), void *ctx)
{
    /* Iterative (sibling/parent links, no recursion, no stack array): a chain
     * can be hundreds of nodes deep and this runs on a small task stack. */
    int cur = g->root, depth = 0;
    while (cur >= 0) {
        cb(ctx, cur, depth);
        if (g->first_child[cur] >= 0) {
            cur = g->first_child[cur];
            depth++;
            continue;
        }
        while (cur != g->root && g->next_sibling[cur] < 0) {
            cur = g->parent[cur];
            depth--;
        }
        if (cur == g->root) {
            break;
        }
        cur = g->next_sibling[cur];
    }
}

/* ── validate ────────────────────────────────────────────────────────────── */

static topo_status_t check_partial(const topo_graph_t *g, const topo_node_t *nodes,
                                   char *reason, size_t reason_len)
{
    size_t n = g->n;
    size_t cap = 0, max_links = 1;
    for (size_t i = 0; i < n; i++) {
        cap += 1 + nodes[i].seen_count;
        if (1 + nodes[i].seen_count > max_links) {
            max_links = 1 + nodes[i].seen_count;
        }
    }
    int *ea      = malloc(cap * sizeof(int));
    int *eb      = malloc(cap * sizeof(int));
    int *parents = malloc(max_links * sizeof(int));
    int *hops    = malloc(n * sizeof(int));
    int *queue   = malloc(n * sizeof(int));
    if (!ea || !eb || !parents || !hops || !queue) {
        free(ea); free(eb); free(parents); free(hops); free(queue);
        snprintf(reason, reason_len, "out of memory checking partial mesh");
        return TOPO_WARN;
    }

    size_t e = 0;
    int multi = 0;
    for (size_t i = 0; i < n; i++) {
        size_t np = 0;
        for (size_t s = 0; s <= nodes[i].seen_count; s++) {
            int p = (s == 0) ? g->parent[i]
                             : resolve(nodes, n, nodes[i].seen_parents + 6 * (s - 1));
            if (p < 0 || p == (int)i) {
                continue;
            }
            bool dup = false;
            for (size_t k = 0; k < np; k++) {
                if (parents[k] == p) {
                    dup = true;
                    break;
                }
            }
            if (dup) {
                continue;
            }
            parents[np++] = p;

            int a = (int)i < p ? (int)i : p;
            int b = (int)i < p ? p : (int)i;
            bool have = false;
            for (size_t k = 0; k < e; k++) {
                if (ea[k] == a && eb[k] == b) {
                    have = true;
                    break;
                }
            }
            if (!have) {
                ea[e] = a;
                eb[e] = b;
                e++;
            }
        }
        if (np >= 2) {
            multi++;
        }
    }

    /* Min-hop layers over the union graph, from the same root. */
    for (size_t i = 0; i < n; i++) {
        hops[i] = 0;
    }
    int head = 0, tail = 0, depth = 0;
    queue[tail++] = g->root;
    hops[g->root] = 1;
    while (head < tail) {
        int u = queue[head++];
        if (hops[u] > depth) {
            depth = hops[u];
        }
        for (size_t k = 0; k < e; k++) {
            int v = (ea[k] == u) ? eb[k] : (eb[k] == u) ? ea[k] : -1;
            if (v >= 0 && hops[v] == 0) {
                hops[v] = hops[u] + 1;
                queue[tail++] = v;
            }
        }
    }
    long vcount = tail;
    long ecount = 0;
    for (size_t k = 0; k < e; k++) {
        if (hops[ea[k]] > 0) {
            ecount++;
        }
    }

    free(ea); free(eb); free(parents); free(hops); free(queue);

    if (vcount >= 3 && ecount == vcount * (vcount - 1) / 2) {
        snprintf(reason, reason_len,
                 "every pair of the %ld nodes is linked - that is a full mesh, not partial",
                 vcount);
        return TOPO_WARN;
    }
    if (multi == 0) {
        snprintf(reason, reason_len,
                 "%ld nodes, each seen under only one parent so far - still a plain tree "
                 "(depth %d)", vcount, g->max_layer);
        return TOPO_WARN;
    }
    snprintf(reason, reason_len,
             "%ld links over %ld nodes, %d node(s) seen under 2+ parents, "
             "tree depth %d, min-hop depth %d",
             ecount, vcount, multi, g->max_layer, depth);
    return TOPO_OK;
}

topo_status_t topo_validate(const topo_graph_t *g, const topo_node_t *nodes,
                            topo_kind_t kind, char *reason, size_t reason_len)
{
    char mac[18];
    topo_status_t st = TOPO_OK;

    if (g->n == 0) {
        snprintf(reason, reason_len, "no nodes");
        return TOPO_WARN;
    }
    if (g->cycle) {
        snprintf(reason, reason_len, "parent links form a cycle - not a valid mesh");
        return TOPO_FAIL;
    }
    if (g->root < 0) {
        snprintf(reason, reason_len, "no root found");
        return TOPO_FAIL;
    }

    switch (kind) {
    case TOPO_KIND_STAR:
        for (size_t i = 0; i < g->n; i++) {
            if (g->layer[i] > 2) {
                mac_str(nodes[i].mac, mac);
                snprintf(reason, reason_len,
                         "%s is at layer %d - a star allows only the center (1) and "
                         "its direct nodes (2)", mac, g->layer[i]);
                return TOPO_FAIL;
            }
        }
        mac_str(nodes[g->root].mac, mac);
        snprintf(reason, reason_len, "center %s + %d direct node(s)",
                 mac, g->reachable - 1);
        break;

    case TOPO_KIND_LINEAR: {
        int leaf = g->root;
        for (size_t i = 0; i < g->n; i++) {
            if (g->layer[i] == 0) {
                continue;
            }
            if (g->child_count[i] > 1) {
                mac_str(nodes[i].mac, mac);
                snprintf(reason, reason_len,
                         "%s has %d children - a linear chain allows 1 (branch)",
                         mac, g->child_count[i]);
                return TOPO_FAIL;
            }
            if (g->layer[i] == g->max_layer) {
                leaf = (int)i;
            }
        }
        char mac2[18];
        mac_str(nodes[g->root].mac, mac);
        mac_str(nodes[leaf].mac, mac2);
        snprintf(reason, reason_len, "chain of %d node(s), %s .. %s",
                 g->reachable, mac, mac2);
        break;
    }

    case TOPO_KIND_TREE:
        if (g->reachable >= 3 && g->max_layer <= 2) {
            snprintf(reason, reason_len,
                     "all %d nodes hang directly off the root - flat like a star (depth 2)",
                     g->reachable);
            st = TOPO_WARN;
        } else {
            snprintf(reason, reason_len, "%d node(s), depth %d",
                     g->reachable, g->max_layer);
        }
        break;

    case TOPO_KIND_PARTIAL:
        st = check_partial(g, nodes, reason, reason_len);
        break;

    default:
        snprintf(reason, reason_len, "unknown topology %d", (int)kind);
        return TOPO_FAIL;
    }

    int detached = (int)g->n - g->reachable;
    if (detached > 0) {
        size_t used = strlen(reason);
        if (used < reason_len) {
            snprintf(reason + used, reason_len - used,
                     "; %d node(s) not attached to the root", detached);
        }
        if (st == TOPO_OK) {
            st = TOPO_WARN;
        }
    }
    return st;
}

const char *topo_status_str(topo_status_t s)
{
    switch (s) {
    case TOPO_OK:   return "OK";
    case TOPO_WARN: return "WARN";
    case TOPO_FAIL: return "FAIL";
    default:        return "?";
    }
}

const char *topo_kind_str(topo_kind_t k)
{
    switch (k) {
    case TOPO_KIND_STAR:    return "STAR";
    case TOPO_KIND_TREE:    return "TREE";
    case TOPO_KIND_LINEAR:  return "LINEAR";
    case TOPO_KIND_PARTIAL: return "PARTIAL";
    default:                return "?";
    }
}

File: ESP32-Environment/tools/topology_graph.py
"""
topology_graph.py - derive a mesh's layers from its parent links and check the
result against the built topology's STRUCTURE.

Nothing here assumes a node count, a layer count or a node ID. Layers come
from a breadth-first walk out of whichever node the links show to be the root
(root = layer 1, the ESP-WIFI-MESH convention). Each topology is judged only by
its defining shape:

  star     every node is a direct child of the center - max layer 2 for any N
  linear   every node has at most one child - a chain of any length
  tree     one root, no cycles - any depth
  partial  a tree at any instant, but nodes are seen under more than one parent
           over time, without every pair being linked (that would be full mesh)

Verdicts follow how each shape is produced: STAR and LINEAR are enforced by the
firmware (max_layer 2 / chain + 1 child), so a violation is FAIL. TREE and
PARTIAL shapes come from physical placement, so a mismatch there is WARN. A
cycle in the parent links is FAIL for every topology - it isn't a valid mesh.

components/mesh_common/src/topology_graph.c applies the same rules live on the
root; keep the two in step. Standard library only.
"""

from __future__ import annotations

import random
from collections import deque

TOPOLOGIES = ("star", "tree", "linear", "partial")
OK, WARN, FAIL = "OK", "WARN", "FAIL"
ROOT_LAYER = 1


class Graph:
    """Parent links resolved into a rooted structure.

    parent[node]   -> parent node, or None
    children[node] -> list of child nodes (input order preserved)
    layer[node]    -> 1 for the root; nodes not reachable from it are absent
    """

    def __init__(self, parent_links):
        self.nodes = list(parent_links)
        known = set(self.nodes)
        self.parent = {}
        self.unresolved = []
        self.cycle = False
        for node, par in parent_links.items():
            if par is not None and par not in known:
                self.unresolved.append(node)
                par = None
            if par == node:
                self.cycle = True
                par = None
            self.parent[node] = par

        self.children = {n: [] for n in self.nodes}
        for node in self.nodes:
            par = self.parent[node]
            if par is not None:
                self.children[par].append(node)

        # A parent chain that never reaches a parentless node is a cycle.
        n = len(self.nodes)
        for node in self.nodes:
            cur, steps = node, 0
            while cur is not None and steps <= n:
                cur = self.parent[cur]
                steps += 1
            if cur is not None:
                self.cycle = True
                break

        # Root = the parentless node that carries the most of the mesh. A node
        # that hasn't joined yet is parentless too but reaches only itself. A
        # node whose parent exists but isn't in the data is NOT a root - it
        # only competes if no genuinely parentless node exists.
        self.root = None
        missing = set(self.unresolved)
        for allow_missing in (False, True):
            best = 0
            for node in self.nodes:
                if self.parent[node] is None and (allow_missing or node not in missing):
                    reach = len(self._bfs(node))
                    if reach > best:
                        self.root, best = node, reach
            if self.root is not None:
                break
        self.layer = self._bfs(self.root) if self.root is not None else {}
        self.max_layer = max(self.layer.values(), default=0)

    def _bfs(self, start):
        layer = {start: ROOT_LAYER}
        queue = deque([start])
        while queue:
            u = queue.popleft()
            for c in self.children[u]:
                if c not in layer:
                    layer[c] = layer[u] + 1
                    queue.append(c)
        return layer

    @property
    def reachable(self):
        return len(self.layer)

    @property
    def detached(self):
        return [n for n in self.nodes if n not in self.layer]

    def leaf(self):
        """Deepest reachable node (the far endpoint of a chain)."""
        return max(self.layer, key=lambda n: self.layer[n]) if self.layer else None


def build_graph(parent_links):
    return Graph(parent_links)


def union_layers(graph, union_edges):
    """Min-hop layers over every link ever observed (undirected), from the
    graph's root. union_edges: iterable of (child, parent) pairs."""
    adj = {n: set() for n in graph.nodes}
    for a, b in union_edges:
        if a in adj and b in adj and a != b:
            adj[a].add(b)
            adj[b].add(a)
    if graph.root is None:
        return {}, adj
    layer = {graph.root: ROOT_LAYER}
    queue = deque([graph.root])
    while queue:
        u = queue.popleft()
        for v in adj[u]:
            if v not in layer:
                layer[v] = layer[u] + 1
                queue.append(v)
    return layer, adj


def validate(topology, graph, union_edges=None):
    """(status, reason) for the graph against the topology's structural rules."""
    if topology not in TOPOLOGIES:
        return FAIL, f"unknown topology {topology!r}"
    if not graph.nodes:
        return WARN, "no nodes"
    if graph.cycle:
        return FAIL, "parent links form a cycle - not a valid mesh"
    if graph.root is None:
        return FAIL, "no root found"

    status = OK
    if topology == "star":
        deep = [n for n, ly in graph.layer.items() if ly > 2]
        if deep:
            n = deep[0]
            return FAIL, (f"{n} is at layer {graph.layer[n]} - a star allows only the "
                          f"center (1) and its direct nodes (2)")
        reason = f"center {graph.root} + {graph.reachable - 1} direct node(s)"

    elif topology == "linear":
        branch = [n for n in graph.layer if len(graph.children[n]) > 1]
        if branch:
            n = branch[0]
            return FAIL, (f"{n} has {len(graph.children[n])} children - a linear "
                          f"chain allows 1 (branch)")
        reason = f"chain of {graph.reachable} node(s), {graph.root} .. {graph.leaf()}"

    elif topology == "tree":
        if graph.reachable >= 3 and graph.max_layer <= 2:
            status = WARN
            reason = (f"all {graph.reachable} nodes hang directly off the root - "
                      f"flat like a star (depth 2)")
        else:
            reason = f"{graph.reachable} node(s), depth {graph.max_layer}"

    else:  # partial
        edges = [(n, graph.parent[n]) for n in graph.nodes if graph.parent[n] is not None]
        edges += list(union_edges or [])
        hops, adj = union_layers(graph, edges)
        parents_seen = {}
        for child, par in edges:
            if child in adj and par in adj and child != par:
                parents_seen.setdefault(child, set()).add(par)
        multi = sum(1 for s in parents_seen.values() if len(s) >= 2)
        v = len(hops)
        e = sum(len(adj[n]) for n in hops) // 2
        if v >= 3 and e == v * (v - 1) // 2:
            status = WARN
            reason = f"every pair of the {v} nodes is linked - that is a full mesh, not partial"
        elif multi == 0:
            status = WARN
            reason = (f"{v} nodes, each seen under only one parent so far - still a "
                      f"plain tree (depth {graph.max_layer})")
        else:
            reason = (f"{e} links over {v} nodes, {multi} node(s) seen under 2+ parents, "
                      f"tree depth {graph.max_layer}, min-hop depth "
                      f"{max(hops.values(), default=0)}")

    detached = graph.detached
    if detached:
        reason += f"; {len(detached)} node(s) not attached to the root"
        if status == OK:
            status = WARN
    return status, reason


# ── synthetic graphs (tests, fake data) ─────────────────────────────────────

def synthetic_roster(topology, n, depth=None, seed=None):
    """Fake-data rows for a valid structure: [(node_id, role, layer, parent_mac)].

    Node 0 is NODE_ROOT01, the rest NODE_VICTIM01.. (the existing fixture
    names). Each node gets a synthetic STA MAC; parent_mac is the parent's
    SoftAP BSSID (STA + 1) the way the firmware logs it, all zeros on the root.
    Layers are derived from the generated links, root = 1.
    """
    links, _ = generate(topology, n, depth=depth, seed=seed)
    graph = build_graph(links)

    def sta(i):
        return 0x020000000000 + i * 4

    def fmt(v):
        return ":".join(f"{(v >> s) & 0xFF:02X}" for s in range(40, -1, -8))

    rows = []
    for i in range(n):
        node_id = "NODE_ROOT01" if i == 0 else f"NODE_VICTIM{i:02d}"
        role = "root" if i == 0 else "victim"
        par = links[i]
        parent_mac = "00:00:00:00:00:00" if par is None else fmt(sta(par) + 1)
        rows.append((node_id, role, graph.layer[i], parent_mac))
    return rows



def generate(topology, n, depth=None, seed=None, alt_parent_fraction=0.34,
             partial_fanout=2):
    """Valid synthetic structure of any size.

    Returns (parent_links, history): parent_links maps node index -> parent
    index (None for the root, node 0); history is a list of extra
    (child, parent) links observed over time - only non-empty for partial.
    """
    if topology not in TOPOLOGIES:
        raise ValueError(f"unknown topology {topology!r}")
    if n < 1:
        raise ValueError("need at least one node")
    rng = random.Random(seed)
    links = {0: None}
    history = []

    if topology == "linear":
        for i in range(1, n):
            links[i] = i - 1

    elif topology == "star":
        for i in range(1, n):
            links[i] = 0

    elif topology == "tree":
        depth = depth if depth is not None else max(3, min(n, 4))
        if depth > n:
            raise ValueError(f"a tree of depth {depth} needs at least {depth} nodes")
        for i in range(1, depth):                 # spine gives the exact depth
            links[i] = i - 1
        layer = {i: i + 1 for i in range(depth)}
        for i in range(depth, n):                 # the rest branch off, never deeper
            par = rng.choice([p for p in layer if layer[p] < depth])
            links[i] = par
            layer[i] = layer[par] + 1

    else:  # partial: bounded fan-out tree + alternate parents seen over time
        layer = {0: 1}
        kids = {0: 0}
        for i in range(1, n):
            par = rng.choice([p for p in layer if kids[p] < partial_fanout])
            links[i] = par
            kids[par] += 1
            kids[i] = 0
            layer[i] = layer[par] + 1
        def alternates(i):
            return [p for p in layer if p != links[i] and p != i and layer[p] < layer[i]]

        for i in range(1, n):
            if rng.random() < alt_parent_fraction and alternates(i):
                history.append((i, rng.choice(alternates(i))))
        if not history:   # a partial mesh needs at least one alternate parent
            for i in range(1, n):
                if alternates(i):
                    history.append((i, rng.choice(alternates(i))))
                    break
    return links, history
